### PR TITLE
Add a loading indicator for PerAccountStore

### LINKED ISSUE #465: Show "Connecting…" banner (or equivalent) when server data is stale
When we lose our real-time stream of events from the server, either because of network failures or because the server has expired the event queue and we need to request a new one:
* #184
* #185

we continue to show the latest server data we had, but it may now be stale. We should communicate to the user that that's what's happening.

In the existing zulip-mobile app, we do this with a banner that reads "Connecting…" and appears just under the app bar.

In this app, we could do the same thing, or we could find a different design. One downside of the "Connecting…" banner is that it occupies some vertical space, which means the rest of the content needs to move down to make room (or else let its top portion be obscured by the banner). Nevertheless it might be the best available option.

Even if we use the same design concept, we'll want to pin down the details: the text style, the layout, the banner's background, the specifics of how it animates in and out.

(In zulip-mobile the details are in `src/common/LoadingBanner.js`; there's no animation. But the conceptually similar "No Internet connection" banner, found in `src/boot/OfflineNoticeProvider.js`, does have an animation… on iOS only, because React Native's implementation breaks on Android. Here on Flutter, we should have no such trouble implementing whatever animation we choose.)

### PR DESCRIPTION
**Pull request by PIG208**
<details>

<summary>Preview</summary>

| isLoading=false | isLoading=true |
| - | - |
| image | image |
| image | image |

</details>

Fixes #465.

### CODE REVIEW

File: lib/model/store.dart [hunk 1]
```diff
@@ -295,6 +295,15 @@ class PerAccountStore extends ChangeNotifier with ChannelStore, MessageStore {
   final GlobalStore _globalStore;
   final ApiConnection connection; // TODO(#135): update zulipFeatureLevel with events
 
+  bool get isLoading => _isLoading;
+  bool _isLoading = false;
+  @visibleForTesting
```
Comment (PIG208) [inline lib/model/store.dart:300]: Useful for a widget test added later.

**Comment by PIG208** [comment]
Oops, just pushed again to fix that.

File: lib/widgets/message_list.dart [hunk 2]
```diff
@@ -255,12 +256,13 @@ class _MessageListPageState extends State<MessageListPage> implements MessageLis
...
-      appBar: AppBar(title: MessageListAppBarTitle(narrow: widget.narrow),
+      appBar: ZulipAppBar(
+        title: MessageListAppBarTitle(narrow: widget.narrow),
+        isLoading: store.isLoading,
         backgroundColor: appBarBackgroundColor,
         shape: removeAppBarBottomBorder
           ? const Border()
-          : null, // i.e., inherit
-      ),
+          : null),
```
Comment (chrisbobbe) [inline lib/widgets/message_list.dart:265]: Let's keep this comment. A reader might see `null` and think it means there's no bottom border. That would be wrong; what `null` actually does is make the app bar inherit its bottom border from the `AppBarTheme`.

**Review by chrisbobbe — COMMENTED** [review]
Thanks!

I'm curious if you explored an approach with a toast: https://github.com/zulip/zulip-flutter/issues/465#issuecomment-1984871814

Would it be easy to put some code in one central place, ensuring that a toast showing the org's name is visible for each account in the loading state? (E.g., for CZO: "Connecting to **Zulip Community**…")

Or, a refinement (that could be annoying to implement): filter that list of toasts down to the account corresponding to the foregrounded screen, and no need to show the org's name.

In the screenshots, I'm noticing the color isn't a great match for the app bar, and the split between the app bar and the sticky recipient header is looking kind of awkward.

If we're going with this linear-indicator approach, let's definitely track that it's missing on the lightbox screen, which needs it because it's a per-account screen too.

**Comment by PIG208** [comment]
Thanks for the review! I didn't experiment with `Toast`. It sounds like another viable option.

Could you elaborate on this?

> Or, a refinement (that could be annoying to implement): filter that list of toasts down to the account corresponding to the foregrounded screen, and no need to show the org's name.

Visually, matching the background color should be fairly simple, because `LinearProgressIndicator` is shipped with the app bar.
```diff
diff --git lib/widgets/app_bar.dart lib/widgets/app_bar.dart
index da49a639..a1c4645b 100644
--- lib/widgets/app_bar.dart
+++ lib/widgets/app_bar.dart
@@ -12,6 +12,6 @@ class ZulipAppBar extends AppBar {
     bottom: PreferredSize(
       preferredSize: const Size.fromHeight(4.0),
       child: (isLoading)
-        ? const LinearProgressIndicator(minHeight: 4.0)
+        ? LinearProgressIndicator(backgroundColor: backgroundColor, minHeight: 4.0)
         : const SizedBox.shrink()));
 }
```

**Comment by PIG208** [comment]
Do we have a toast library in mind to use? Or are we writing one from scratch?

**Comment by gnprice** [comment]
Chat thread about what type of indicator to use:
https://chat.zulip.org/#narrow/stream/48-mobile/topic/banner.2Fsnack.20bar.2Fprogress.20indicator.20for.20connection.20error/near/1909308

**Comment by gnprice** [comment]
> If we're going with this linear-indicator approach, let's definitely track that it's missing on the lightbox screen, which needs it because it's a per-account screen too.

I think it's actually fine to not have the indicator on the lightbox screen. Critically, there's very little there to get live-updated — even if you're looking at stale data there and you think it's up to date, it's hard to get misled about much as a result. In particular it never gets updated by a new message arriving.

It should definitely get a comment saying we've chosen not to put the indicator there, though, and why.

**Comment by chrisbobbe** [comment]
Sure, I guess that all makes sense.

**Comment by chrisbobbe** [comment]
So the current revision is close to what we want, right? Is this the only thing missing:

> It should definitely get a comment saying we've chosen not to put the indicator there, though, and why.

**Comment by PIG208** [comment]
Right, will update this later.

**Comment by PIG208** [comment]
Updated the PR to mention the ZulipAppBar exceptions in comments after its dart doc.

**Comment by PIG208** [comment]
Updated the PR to handle `isLoading` when there are transient connection errors.

**Comment by chrisbobbe** [comment]
Thanks! LGTM, over to Greg for his review.

**Comment by PIG208** [comment]
(Rebased the PR)

File: lib/widgets/app.dart [hunk 3]
```diff
@@ -179,9 +180,10 @@ class ChooseAccountPage extends StatelessWidget {
     assert(!PerAccountStoreWidget.debugExistsOf(context));
     final globalStore = GlobalStoreWidget.of(context);
     return Scaffold(
-      appBar: AppBar(
+      appBar: ZulipAppBar(
         title: Text(zulipLocalizations.chooseAccountPageTitle),
-        actions: const [ChooseAccountPageOverflowButton()]),
+        actions: const [ChooseAccountPageOverflowButton()],
+        isLoading: false),
```
Comment (gnprice) [inline lib/widgets/app.dart]: How about just `AppBar`, if we're going to hard-code `isLoading` here to false?

File: lib/widgets/inbox.dart [hunk 4]
```diff
@@ -160,7 +161,9 @@ class _InboxPageState extends State<InboxPage> with PerAccountStoreAwareStateMix
     }
 
     return Scaffold(
-      appBar: AppBar(title: const Text('Inbox')),
+      appBar: ZulipAppBar(
+        title: const Text('Inbox'),
+        isLoading: store.isLoading),
```
Comment (gnprice) [inline lib/widgets/inbox.dart]: I think this pattern can be simplified. I'll push a separate commit on top to do so.

**Review by gnprice — COMMENTED** [review]
Thanks @PIG208, and thanks @chrisbobbe for the previous reviews!

Generally this looks good. I also put a build with this PR (among others) onto my phone last week before going on vacation, and the behavior was very helpful.

A couple of small comments below. I've also pushed some additional commits on top:
c384847c1 wip squash? plain AppBar where isLoading always false
667adf856 app_bar [nfc]: Move ZulipAppBar.isLoading to within widget
a6965aad5 app_bar [nfc]: Explain why the lightbox skips ZulipAppBar

The first one reflects one comment, and can be squashed; I added it just because that change is needed for the next commit.

**Comment by PIG208** [comment]
Thanks! Rebased the PR and squashed the commit.

**Comment by gnprice** [comment]
Looks good! Thanks for the revision — merging.

**Pull request merged**